File: app/src/main/java/com/quran/labs/androidquran/presenter/QuranImportPresenter.java
package com.quran.labs.androidquran.presenter;

import android.Manifest;
import android.content.Context;
import android.content.Intent;
import android.content.pm.PackageManager;
import android.net.Uri;
import android.os.ParcelFileDescriptor;
import android.support.annotation.NonNull;
import android.support.annotation.VisibleForTesting;
import android.support.v4.app.ActivityCompat;

import com.quran.labs.androidquran.QuranImportActivity;
import com.quran.labs.androidquran.dao.BookmarkData;
import com.quran.labs.androidquran.model.bookmark.BookmarkImportExportModel;
import com.quran.labs.androidquran.model.bookmark.BookmarkModel;
import com.quran.labs.androidquran.service.util.PermissionUtil;
import com.quran.labs.androidquran.util.QuranSettings;

import java.io.FileDescriptor;
import java.io.FileInputStream;
import java.io.InputStream;
import java.util.concurrent.Callable;

import javax.inject.Inject;
import javax.inject.Singleton;

import io.reactivex.Maybe;
import io.reactivex.MaybeSource;
import io.reactivex.Observable;
import io.reactivex.android.schedulers.AndroidSchedulers;
import io.reactivex.functions.Function;
import io.reactivex.observers.DisposableMaybeObserver;
import io.reactivex.schedulers.Schedulers;
import okio.BufferedSource;
import okio.Okio;


@Singleton
public class QuranImportPresenter implements Presenter<QuranImportActivity> {
  private static final int REQUEST_WRITE_TO_SDCARD_PERMISSIONS = 1;

  private final Context mAppContext;
  private final BookmarkModel mBookmarkModel;
  private final BookmarkImportExportModel mBookmarkImportExportModel;

  private boolean mRequestingPermissions;
  private Observable<Boolean> mImportObservable;
  private QuranImportActivity mCurrentActivity;

  @Inject
  QuranImportPresenter(Context appContext,
                       BookmarkImportExportModel model,
                       BookmarkModel bookmarkModel) {
    mAppContext = appContext;
    mBookmarkModel = bookmarkModel;
    mBookmarkImportExportModel = model;
  }

  private void handleIntent(Intent intent) {
    mRequestingPermissions = false;
    if (mImportObservable == null) {
      Uri uri = intent.getData();
      if (uri == null && intent.getExtras() != null) {
        uri = (Uri) intent.getExtras().get(Intent.EXTRA_STREAM);
      }

      if (uri != null) {
        parseIntentUri(uri);
      } else if (mCurrentActivity != null) {
        mCurrentActivity.showError();
      }
    } else {
      subscribeToImportData();
    }
  }

  public void importData(final BookmarkData data) {
    mImportObservable = mBookmarkModel.importBookmarksObservable(data);
    subscribeToImportData();
  }

  public void onPermissionsResult(int requestCode, @NonNull int[] grantResults) {
    if (requestCode == REQUEST_WRITE_TO_SDCARD_PERMISSIONS) {
      mRequestingPermissions = false;
      if (mCurrentActivity != null) {
        if (grantResults.length == 1 && grantResults[0] == PackageManager.PERMISSION_GRANTED) {
          handleIntent(mCurrentActivity.getIntent());
        } else {
          mCurrentActivity.showPermissionsError();
        }
      }
    }
  }

  private void subscribeToImportData() {
    mImportObservable
        .observeOn(AndroidSchedulers.mainThread())
        .subscribe(aBoolean -> {
          if (mCurrentActivity != null) {
            mCurrentActivity.showImportComplete();
            mImportObservable = null;
          }
        });
  }

  private void parseIntentUri(final Uri uri) {
    getBookmarkDataObservable(parseUri(uri))
        .observeOn(AndroidSchedulers.mainThread())
        .subscribe(new DisposableMaybeObserver<BookmarkData>() {
          @Override
          public void onSuccess(BookmarkData bookmarkData) {
            if (mCurrentActivity != null) {
              mCurrentActivity.showImportConfirmationDialog(bookmarkData);
            }
          }

          @Override
          public void onError(Throwable e) {
            if (mCurrentActivity != null) {
              handleExternalStorageFile(uri);
            }
          }

          @Override
          public void onComplete() {
            if (mCurrentActivity != null) {
              handleExternalStorageFile(uri);
            }
          }
        });
  }

  private void handleExternalStorageFile(Uri uri) {
    if (PermissionUtil.haveWriteExternalStoragePermission(mAppContext)) {
      handleExternalStorageFileInternal(uri);
    } else if (mCurrentActivity != null) {
      mRequestingPermissions = true;
      if (PermissionUtil.canRequestWriteExternalStoragePermission(mCurrentActivity)) {
        QuranSettings.getInstance(mAppContext).setSdcardPermissionsDialogPresented();
        ActivityCompat.requestPermissions(mCurrentActivity,
            new String[]{ Manifest.permission.WRITE_EXTERNAL_STORAGE },
            REQUEST_WRITE_TO_SDCARD_PERMISSIONS);
      } else {
        mCurrentActivity.showPermissionsError();
      }
    }
  }

  private void handleExternalStorageFileInternal(Uri uri) {
    getBookmarkDataObservable(parseExternalFile(uri))
        .observeOn(AndroidSchedulers.mainThread())
        .subscribe(new DisposableMaybeObserver<BookmarkData>() {
          @Override
          public void onSuccess(BookmarkData bookmarkData) {
            if (mCurrentActivity != null) {
              mCurrentActivity.showImportConfirmationDialog(bookmarkData);
            }
          }

          @Override
          public void onError(Throwable e) {
            if (mCurrentActivity != null) {
              mCurrentActivity.showError();
            }
          }

          @Override
          public void onComplete() {
            if (mCurrentActivity != null) {
              mCurrentActivity.showError();
            }
          }
        });
  }

  private Maybe<BookmarkData> getBookmarkDataObservable(Maybe<BufferedSource> source) {
    return source
        .flatMap(new Function<BufferedSource, MaybeSource<BookmarkData>>() {
          @Override
          public MaybeSource<BookmarkData> apply(BufferedSource bufferedSource) throws Exception {
            return mBookmarkImportExportModel.readBookmarks(bufferedSource).toMaybe();
          }
        })
        .subscribeOn(Schedulers.io());
  }

  @NonNull
  @VisibleForTesting(otherwise = VisibleForTesting.PRIVATE)
  Maybe<BufferedSource> parseUri(final Uri uri) {
    return Maybe.defer(new Callable<MaybeSource<BufferedSource>>() {
      @Override
      public MaybeSource<BufferedSource> call() throws Exception {
        ParcelFileDescriptor pfd = mAppContext.getContentResolver().openFileDescriptor(uri, "r");
        if (pfd != null) {
          FileDescriptor fd = pfd.getFileDescriptor();
          return Maybe.just(Okio.buffer(Okio.source(new FileInputStream(fd))));
        }
        return Maybe.empty();
      }
    });
  }

  @NonNull
  @VisibleForTesting(otherwise = VisibleForTesting.PRIVATE)
  Maybe<BufferedSource> parseExternalFile(final Uri uri) {
    return Maybe.defer(new Callable<MaybeSource<BufferedSource>>() {
      @Override
      public MaybeSource<BufferedSource> call() throws Exception {
        InputStream stream = mAppContext.getContentResolver().openInputStream(uri);
        if (stream != null) {
          return Maybe.just(Okio.buffer(Okio.source(stream)));
        }
        return Maybe.empty();
      }
    });
  }

  @Override
  public void bind(QuranImportActivity activity) {
    mCurrentActivity = activity;
    if (!activity.isShowingDialog() && !mRequestingPermissions) {
      handleIntent(activity.getIntent());
    }
  }

  @Override
  public void unbind(QuranImportActivity activity) {
    if (activity == mCurrentActivity) {
      mCurrentActivity = null;
    }
  }
}

File: app/src/main/java/com/quran/labs/androidquran/presenter/QuranImportPresenter.kt
package com.quran.labs.androidquran.presenter

import android.Manifest
import android.content.Context
import android.content.Intent
import android.content.pm.PackageManager
import android.net.Uri
import androidx.annotation.VisibleForTesting
import androidx.core.app.ActivityCompat
import com.quran.data.model.bookmark.BookmarkData
import com.quran.labs.androidquran.QuranImportActivity
import com.quran.labs.androidquran.model.bookmark.BookmarkImportExportModel
import com.quran.labs.androidquran.model.bookmark.BookmarkModel
import com.quran.labs.androidquran.service.util.PermissionUtil.canRequestWriteExternalStoragePermission
import com.quran.labs.androidquran.service.util.PermissionUtil.haveWriteExternalStoragePermission
import com.quran.labs.androidquran.util.QuranSettings
import com.quran.mobile.di.qualifier.ApplicationContext
import io.reactivex.rxjava3.android.schedulers.AndroidSchedulers
import io.reactivex.rxjava3.core.Maybe
import io.reactivex.rxjava3.core.Observable
import io.reactivex.rxjava3.disposables.CompositeDisposable
import io.reactivex.rxjava3.observers.DisposableMaybeObserver
import io.reactivex.rxjava3.schedulers.Schedulers
import okio.BufferedSource
import okio.buffer
import okio.source
import java.io.FileInputStream
import javax.inject.Inject
import javax.inject.Singleton

@Singleton
class QuranImportPresenter @Inject internal constructor(
  @ApplicationContext private val appContext: Context,
  private val bookmarkImportExportModel: BookmarkImportExportModel,
  private val bookmarkModel: BookmarkModel
) : Presenter<QuranImportActivity> {
  private val compositeDisposable = CompositeDisposable()

  private var requestingPermissions = false
  private var importObservable: Observable<Boolean>? = null
  private var currentActivity: QuranImportActivity? = null

  private fun handleIntent(intent: Intent) {
    requestingPermissions = false
    if (importObservable == null) {
      val uri = intent.data ?: intent.extras?.get(Intent.EXTRA_STREAM) as Uri?
      if (uri != null) {
        parseIntentUri(uri)
      } else {
        currentActivity?.showError()
      }
    } else {
      subscribeToImportData()
    }
  }

  fun importData(data: BookmarkData) {
    importObservable = bookmarkModel.importBookmarksObservable(data)
    subscribeToImportData()
  }

  fun onPermissionsResult(requestCode: Int, grantResults: IntArray) {
    if (requestCode == REQUEST_WRITE_TO_SDCARD_PERMISSIONS) {
      requestingPermissions = false
      val currentActivity = currentActivity
      if (currentActivity != null) {
        if (grantResults.size == 1 && grantResults[0] == PackageManager.PERMISSION_GRANTED) {
          handleIntent(currentActivity.intent)
        } else {
          currentActivity.showPermissionsError()
        }
      }
    }
  }

  private fun subscribeToImportData() {
    val observable = importObservable
    if (observable != null) {
      compositeDisposable.add(
        observable
          .observeOn(AndroidSchedulers.mainThread())
          .subscribe { ignored: Boolean? ->
            currentActivity?.showImportComplete()
            importObservable = null
          })
    }
  }

  private fun parseIntentUri(uri: Uri) {
    getBookmarkDataObservable(parseUri(uri))
      .observeOn(AndroidSchedulers.mainThread())
      .subscribe(object : DisposableMaybeObserver<BookmarkData>() {
        override fun onSuccess(bookmarkData: BookmarkData) {
          currentActivity?.showImportConfirmationDialog(bookmarkData)
        }

        override fun onError(e: Throwable) {
          if (currentActivity != null) {
            handleExternalStorageFile(uri)
          }
        }

        override fun onComplete() {
          if (currentActivity != null) {
            handleExternalStorageFile(uri)
          }
        }
      })
  }

  private fun handleExternalStorageFile(uri: Uri) {
    val currentActivity = currentActivity
    if (haveWriteExternalStoragePermission(appContext)) {
      handleExternalStorageFileInternal(uri)
    } else if (currentActivity != null) {
      requestingPermissions = true
      if (canRequestWriteExternalStoragePermission(currentActivity)) {
        QuranSettings.getInstance(appContext).setSdcardPermissionsDialogPresented()
        ActivityCompat.requestPermissions(
          currentActivity,
          arrayOf(Manifest.permission.WRITE_EXTERNAL_STORAGE),
          REQUEST_WRITE_TO_SDCARD_PERMISSIONS
        )
      } else {
        currentActivity.showPermissionsError()
      }
    }
  }

  private fun handleExternalStorageFileInternal(uri: Uri) {
    getBookmarkDataObservable(parseExternalFile(uri))
      .observeOn(AndroidSchedulers.mainThread())
      .subscribe(object : DisposableMaybeObserver<BookmarkData>() {
        override fun onSuccess(bookmarkData: BookmarkData) {
          currentActivity?.showImportConfirmationDialog(bookmarkData)
        }

        override fun onError(e: Throwable) {
          currentActivity?.showError()
        }

        override fun onComplete() {
          currentActivity?.showError()
        }
      })
  }

  private fun getBookmarkDataObservable(source: Maybe<BufferedSource>): Maybe<BookmarkData> {
    return source
      .flatMap { bufferedSource: BufferedSource ->
        bookmarkImportExportModel.readBookmarks(bufferedSource).toMaybe()
      }
      .subscribeOn(Schedulers.io())
  }

  @VisibleForTesting
  fun parseUri(uri: Uri): Maybe<BufferedSource> {
    return Maybe.defer {
      val pfd = appContext.contentResolver.openFileDescriptor(uri, "r")
      if (pfd != null) {
        val fd = pfd.fileDescriptor
        return@defer Maybe.just<BufferedSource>(
          FileInputStream(
            fd
          ).source().buffer()
        )
      }
      Maybe.empty()
    }
  }

  @VisibleForTesting
  fun parseExternalFile(uri: Uri): Maybe<BufferedSource> {
    return Maybe.defer {
      val stream = appContext.contentResolver.openInputStream(uri)
      if (stream != null) {
        return@defer Maybe.just<BufferedSource>(
          stream.source().buffer()
        )
      }
      Maybe.empty()
    }
  }

  override fun bind(what: QuranImportActivity) {
    currentActivity = what
    if (!what.isShowingDialog() && !requestingPermissions) {
      handleIntent(what.intent)
    }
  }

  override fun unbind(what: QuranImportActivity) {
    if (what == currentActivity) {
      currentActivity = null
      compositeDisposable.clear()
    }
  }

  companion object {
    private const val REQUEST_WRITE_TO_SDCARD_PERMISSIONS = 1
  }
}
